libraries/chain/include/eosio/chain/fork_database.hpp

```diff
@@ -34,9 +34,10 @@ namespace eosio::chain {
       using bhsp             = bs::bhsp_t;
       using bhs              = bhsp::element_type;
       using bsp_t            = bsp;
-      using branch_type      = deque<bsp>;
+      using branch_type      = std::vector<bsp>;
+      using full_branch_type = std::vector<bhsp>;
```

Comment:
Is the idea that we should use vector because these are expected to be small? That probably makes sense, but curious why the change.

Reply:
Sure, since they are expected to be small, I think vector is more appropriate for sure. 

But even for hundreds  or thousands typically I think we should prefer vectors to `deque`, because we can `reserve` the appropriate size (fewer memory allocations) and subsequent access is cache friendly because memory is contiguous.

IMO we significantly overuse `deque` in our code base.

Reply:
If you don't agree I don't mind changing `branch_type` back to `deque`.

Reply:
I think vector is fine for these `shared_ptr`s as the size is small for each one and our fork db should never be very deep. Even in dpos it is typically only about 12.
As for `significantly overuse deque in our code base`, we demonstrated significate performance improvement by using `deque` instead of `vector` in controller. Those are also for small types, although the size was ~10K. ~20K TPS.
On other projects, I have also seen issues with `vector` not providing the performance benefit often discussed.

Reply:
deque is less like list and more like vector. You get groups of locality of reference unlike wth list.

Reply:
Yes, indeed, but the size of the group is implementation-defined and unreliable, as we see with the Microsoft STL version. What benefit do you see with `deque` in this use case?

Reply:
I don't see any in this use-case. Our branches should be very small.

Reply:
Our `deque` is defined as:
```
   using block_1024_option_t = boost::container::deque_options< boost::container::block_size<1024u> >::type;
   template<typename T>
   using deque = boost::container::deque< T, void, block_1024_option_t >;
```

Reply:
Yes, using this deque definition will prevent the pathological poor behavior of Microsoft's version.
Still, it seems to me that a  `1024` block size is, more often than not, less optimal than using a correctly `reserve'd` vector.

Reply:
When you know the size ahead of time, vector is clear winner in all cases. It is when you don't know and you vector has to grow multiple times that it can kill performance.